warehouse/forklift/legacy.py

```diff
@@ -1439,7 +1474,7 @@ def file_upload(request):
                 action="add {python_version} file {filename}".format(
                     python_version=file_.python_version, filename=file_.filename
                 ),
-                submitted_by=request.user,
+                submitted_by=request.user if request.user else None,
```

Comment:
Yeah, I think that's a good idea. Two items from this:

* We need some way to get the `OIDCProvider` in this context. `request.identity` isn't good enough (since it's just the `Project`) and our data model currently allows a project to have multiple providers. Maybe we eliminate the many-many relationship there so that we can reliably get `project.oidc_provider` when `isinstance(identity, Project)`?
* We'll need to retain `submitted_by` as non-optional, but drop the `User` relationship. It's currently a `CIText`, so maybe we have to change the column type as well?

Reply:
I think we need to retain the many:many, especially for the future when we have multiple providers, and a project might need to use multiple providers to publish.

We probably need a way to retain the publisher used to authorize the request. We could possibly use `request.authentication_method` from https://github.com/pypa/warehouse/pull/10854 or something similar for this, getting these details from the macaroon (re my previous comment about adding this information to the macaroon).

Reply:
Gotcha. I think the macaroon is probably the right place for this -- I can add something like an `OIDCPublisherCaveat` that stores the DB ID for the publisher, which we can then sidecar somewhere in the `request`.

Reply:
I took another look at this, and here's how `submitted_by` is defined and related:

```py
    _submitted_by = Column(
        "submitted_by",
        CIText,
        ForeignKey("users.username", onupdate="CASCADE"),
        nullable=True,
    )
    submitted_by = orm.relationship(User, lazy="raise_on_sql")
```

So we have a couple of options:

* Remove the FK constraint + relationship and turn `submitted_by` into an ordinary `CIText` column
* Add a separate column + relationship for `JournalEvent`s submitted by `OIDCProvider`s, additionally constraining them such that `JournalEvent -> User XOR OIDCProvider` (similar to the `Macaroon` table)
* More extremely: journal events have been deprecated for a while (since 2019), so maybe we can just delete this entirely? 

Thoughts?

Reply:
Sounds good, I'll add that.

Reply:
I'll do this as part of #12477 as well.